File: objc/Project8/Project8/GameViewController.m
//
//  GameViewController.m
//  Project8
//
//  Created by Jinwoo Kim on 2/12/21.
//

#import "GameViewController.h"
#import "GameScene.h"
#import "PlayScene.h"

@implementation GameViewController

- (SKScene * _Nullable)gameScene {
    return ((SKView *)self.view).scene;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Load the SKScene from 'GameScene.sks'
    GameScene *scene = (GameScene *)[SKScene nodeWithFileNamed:@"GameScene"];
    
    // Set the scale mode to scale to fit the window
    scene.scaleMode = SKSceneScaleModeAspectFill;
    
    SKView *skView = (SKView *)self.view;
    
    // Present the scene
    [skView presentScene:scene];
    
    skView.showsFPS = YES;
    skView.showsNodeCount = YES;
    
    UIPanGestureRecognizer *recognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(movementMade:)];
    [self.view addGestureRecognizer:recognizer];
}

- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event {
    UIPress *press = [[presses allObjects] firstObject];
    if (press == nil) return;
    
    if (press.type == UIPressTypeMenu) {
        if ([[self gameScene] isKindOfClass:[PlayScene class]] && ([self gameScene] != nil)) {
            [(PlayScene *)[self gameScene] quitGame];
        } else {
            [super pressesBegan:presses withEvent:event];
        }
    }
    
    [[self gameScene] pressesBegan:presses withEvent:event];
}

- (void)movementMade:(UIPanGestureRecognizer *)recognizer {
    CGPoint translation = [recognizer translationInView:recognizer.view];
    [recognizer setTranslation:CGPointZero inView:recognizer.view];
    
    PlayScene *scene = (PlayScene *)[self gameScene];
    if (scene) [scene movePlayer:translation];
}

@end

File: swift/Project8/Project8/GameViewController.swift
//
//  GameViewController.swift
//  Project8
//
//  Created by Jinwoo Kim on 2/12/21.
//

import UIKit
import SpriteKit
import GameplayKit

class GameViewController: UIViewController {
    
    var gameScene: SKScene? {
        return (self.view as! SKView).scene
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let view = self.view as! SKView? {
            // Load the SKScene from 'GameScene.sks'
            if let scene = SKScene(fileNamed: "GameScene") {
                // Set the scale mode to scale to fit the window
                scene.scaleMode = .aspectFill
                
                // Present the scene
                view.presentScene(scene)
            }
            
            view.ignoresSiblingOrder = true
            
            view.showsFPS = true
            view.showsNodeCount = true
            
            let recognizer = UIPanGestureRecognizer(target: self, action: #selector(movementMade))
            view.addGestureRecognizer(recognizer)
        }
    }

    override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
        guard let press = presses.first else { return }
        
        if press.type == .menu {
            if let scene = gameScene as? PlayScene {
                scene.quitGame()
            } else {
                super.pressesBegan(presses, with: event)
            }
        }
        
        gameScene?.pressesBegan(presses, with: event)
    }
    
    @objc func movementMade(recognizer: UIPanGestureRecognizer) {
        let translation = recognizer.translation(in: recognizer.view)
        recognizer.setTranslation(.zero, in: recognizer.view)
        
        if let scene = gameScene as? PlayScene {
            scene.movePlayer(translation)
        }
    }
}
